include/openthread/tcp.h

```diff
@@ -0,0 +1,616 @@
+/**
+ * This callback indicates the number of bytes available for consumption from
+ * the receive buffer. It is called whenever bytes are added to the receive
+ * buffer. If the end of the stream has been reached (i.e., if no more data
+ * will become available to read because the connection peer has closed their
+ * end of the connection for writing), then @p aEndOfStream is true. Finally,
+ * @p aBytesRemaining indicates how much capacity is left in the receive
+ * buffer to hold additional data that arrives.
+ */
+typedef void (*otTcpReceiveAvailable)(void *aContext, otTcpEndpoint *aEndpoint, size_t aBytesAvailable, bool aEndOfStream, size_t aBytesRemaining);
+
+typedef enum otTcpDisconnectedReason {
+    OT_TCP_DISCONNECTED_REASON_NORMAL,
+    OT_TCP_DISCONNECTED_REASON_REFUSED,
+    OT_TCP_DISCONNECTED_REASON_RESET,
+    OT_TCP_DISCONNECTED_REASON_TIME_WAIT,
+    OT_TCP_DISCONNECTED_REASON_TIMED_OUT,
+} otTcpDisconnectedReason;
+
+/**
+ * This callback indicates that the connection was broken and should no longer
+ * be used, or that a connection has entered the TIME-WAIT state. It can occur
+ * if a connection establishment attempt (initiated by calling @p otTcpConnect)
+ * fails, or any point thereafter (e.g., if the connection times out or an RST
+ * segment is received from the connection peer. Once this callback fires, all
+ * resources that the application provided for this connection (i.e., any
+ * `otLinkedBuffers` and memory they reference, but not the TCP endpoint itself
+ * or space for the receive buffers) can be reclaimed. In the case of a
+ * connection entering the TIME-WAIT state, this callback is called twice, once
+ * upon entry into the TIME-WAIT state (with
+ * OT_TCP_DISCONNECTED_REASON_TIME_WAIT, and again when the TIME-WAIT state
+ * expires (with OT_TCP_DISCONNECTED_REASON_NORMAL).
+ */
+typedef void (*otTcpDisconnected)(void *aContext, otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason);
+
+/**
+ * This structure represents a TCP endpoint.
+ *
+ * An TCP endpoint acts an endpoint of TCP connection. It can be used to
+ * initiate TCP connections, and, once a TCP connection is established, send
+ * data to and receive data from the connection peer.
+ */
+typedef struct otTcpEndpoint
+{
+    struct otTcpEndpoint *mNext;                      ///< A pointer to the next TCP endpoint (internal use only)
+    void *mContext;                                   ///< A pointer to application-specific context
+    otTcpEstablished mEstablishedCallback;            ///< "Established" callback function
+    otTcpSendDone mSendDoneCallback;                  ///< "Send done" callback function
+    otTcpSendReady mSendReadyCallback;                ///< "Send ready" callback function
+    otTcpReceiveAvailable mReceiveAvailableCallback;  ///< "Receive available" callback function
+    otTcpDisconnected mDisconnectedCallback;          ///< "Disconnected" callback function
+
+    /* Other implementation-defined fields go here. */
```

Comment:
Thinking a bit more on this, I think it would be nicer if we can somehow hide the members of `otTcpEndpoint` (define it an opaque type) while still having the application provide the space/buffer for it.

How about the following:
- Application is expected to give us a buffer (`mReceiveBuffer` in `otTcpEndpointInitializeArgs`) to use with the end pint.
- We can try to allocate a `otTcpEndpoint` from the same buffer (and require the app to give us a lager buffer)
- Basically we change `otTcpEndpointInitialize()` to return a `otTcpEndpoint *` back to the caller.

With this, we can extend the internal definition of `otTcpEndPoint` struct (add/remove fields) without the need for public APIs and definitions to change.

Reply:
Afterwards I realized we can only do such a model with `otTcpEndpoint` and there would no similar way for `otTcpListener` struct. So I agree it may be better to use the same model for both `otTcpEndPpoint` and `otTcpListener` structs and have caller provide the `struct`.

Reply:
I added documentation on the structs making it clear to the user that `otTcpEndpoint`'s and `otTcpListener`'s fields should not be accessed directly.